README.md
# Vydumschik (Выдумщик) [![Travis CI status](https://secure.travis-ci.org/leonid-shevtsov/vydumschik.png)](http://travis-ci.org/leonid-shevtsov/vydumschik)

This gem generates fake data for database population in Russian. It probably won't be of any use to you if you don't know the language.

`Vydumschik`- гем, выдумывающий случайные данные для наполнения баз данных. Вдохновлен гемом [`ffaker`](http://rubygems.org/gems/ffaker).

## Установка

Зависимостей гем не имеет - но **(!)** - работает исключительно с кодировкой UTF-8. При использовании Ruby 1.9 создаваемые строки можно легко и незаметно преобразовывать в любую другую кодировку.

    gem install vydumschik

## Имена

    > require 'vydumschik'

    > Vydumschik::Name.first_name
    "Филипп"

    > Vydumschik::Name.middle_name(:female)
    "Артёмовна"

    > Vydumschik::Name.surname
    "Шарапов"

    > Vydumschik::Name.full_name
    "Королёва Устиния Панкратиевна"

## Адреса

    > Vydumschik::Address.street
    "ул. Парниковая"

    > Vydumschik::Address.street_address
    "ул. Линейная, д. 50, кв. 9"

    > Vydumschik::Address.street_address
    "ул. Нефтяников, 36/14"

    > Vydumschik::Address.street_address
    "ул. Войкова, д. 1"

## Лорем ипсум

    > Vydumschik::Lorem.word
    "жиганир"

    > Vydumschik::Lorem.sentence
    "Инекуле тято авитудаз еде агар."

    > Vydumschik::Lorem.paragraph
    "Орэнет емех рахетик копорун хетег динава. Оте балале утир лелунака азуло
    илувап. Татиройы отедечу ыжа омотокях мизота. Исяр одокыдика оненен итотип
    осякелоза ув онежорес ресовинон. Уто ате обю иро авушытяса енесета ран
    конекесуг омат. Хованово диномиг лодер ивен косе еламалип са тин. Ойон
    анаро совизоцюч деке есе дажочи. Екечовено лета рак оня онал."


(c) 2011 [Leonid Shevtsov](http://leonid.shevtsov.me)


CHANGELOG.md
## 2012-10-15 0.2

* Ensured compatibility with Ruby 1.9

spec/name_spec.rb
require 'spec_helper'

describe Vydumschik::Name do
  subject { described_class }

  it "should generate first names" do
    subject.first_name.should be_a String
  end

  it "should generate middle names" do
    subject.middle_name.should be_a String
  end

  it "should generate surnames" do
    subject.surname.should be_a String
  end

  it "should generate full names" do
    subject.full_name.should be_a String
  end
end

spec/lorem_spec.rb
require 'spec_helper'

describe Vydumschik::Lorem do
  subject { described_class }

  it "should generate words" do
    subject.word.should be_a String
  end

  it "should generate sentences" do
    subject.sentence.should be_a String
  end

  it "should generate paragraphs" do
    subject.paragraph.should be_a String
  end
end

spec/spec_helper.rb
require 'vydumschik'

lib/vydumschik/name.rb
# encoding: utf-8

require 'yaml'

module Vydumschik

  # This module generates human names
  module Name

    # Random first name (gender = :male/:female/nil for random)
    def self.first_name(gender=nil)
      gender ||= random_gender
      gender == :female ? data[:female][rand data[:female].length] : data[:male][rand data[:male].length][:name]
    end

    # Random middle name (gender = :male/:female/nil for random)
    def self.middle_name(gender=nil)
      gender ||= random_gender
      name = data[:male][rand data[:male].length]
      gender == :female ? name[:female_middle] : name[:male_middle]
    end

    # Random surnamename (gender = :male/:female/nil for random)
    def self.surname(gender=nil)
      gender ||= random_gender
      surname = data[:surnames][rand data[:surnames].length]
      surname += 'а' if gender==:female
      surname
    end

    # Random full name - surname, first name and middle name (gender = :male/:female/nil for random)
    def self.full_name(gender=nil)
      gender ||= random_gender
      [surname(gender), first_name(gender), middle_name(gender)].join(' ')
    end

    private

    def self.data
      unless @data
        @data = YAML.load_file(File.expand_path('../../../data/names.yml', __FILE__))
        encode_data_to_utf8 if RUBY_VERSION >= '1.9'
      end
      @data
    end

    def self.random_gender
      rand>0.5 ? :female : :male
    end

    def self.encode_data_to_utf8
      @data[:male].each do |s|
        s[:male_middle].force_encoding('utf-8')
        s[:female_middle].force_encoding('utf-8')
        s[:name].force_encoding('utf-8')
      end
      @data[:female].each {|s| s.force_encoding('utf-8') }
      @data[:surnames].each {|s| s.force_encoding('utf-8') }
    end
  end
end


lib/vydumschik/lorem.rb
# encoding: utf-8


module Vydumschik

  # This module generate lipsum-like russian pseudosentences.
  #
  # Warning!
  # To avoid using ActiveSupport for multibyte support, or anything similar,
  # I just assume that all russian chars are 2-byte (as in UTF8) and that I
  # always get a russian char when I ask for one.
  module Lorem

    # Word length
    WORD_LIMITS = [2, 10]

    # Sentence length, in words
    SENTENCE_LIMITS = [5, 10]

    # Paragraph length, in sentences
    PARA_LIMITS = [5, 10]

    # Russian consonants
    CONSONANTS = %w(б в г д ж з й к л м н п р с т ф х ц ч ш щ)

    # Russian vowels
    VOWELS = %w(а е ё и о у ы э ю я)

    # Consonants, normalized by character frequency
    CONSONANTS_NORMALIZED = %w(б б б б б б б б б б б б б б б б б б б в в в в в в в в в в в в в в в в в в в в в в в в в в в в в в в в в в в в в в в в в в в в в в в в г г г г г г г г г г г г г г г г г г г д д д д д д д д д д д д д д д д д д д д д д д д д д д д д д д д ж ж ж ж ж ж ж ж ж ж ж ж з з з з з з з з з з з з з з з з з з й й й й й й й й й й й к к к к к к к к к к к к к к к к к к к к к к к к к к к к к к к к к к к к л л л л л л л л л л л л л л л л л л л л л л л л л л л л л л л л л л л л л л л л л л л л л л л л л л л л л л м м м м м м м м м м м м м м м м м м м м м м м м м м м м м м м м н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н н п п п п п п п п п п п п п п п п п п п п п п п п п п р р р р р р р р р р р р р р р р р р р р р р р р р р р р р р р р р р р р р р р р р р р р р с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с с т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т т ф х х х х х х х х ц ц ц ч ч ч ч ч ч ч ч ч ч ч ч ч ч ч ч ч ч ш ш ш ш ш ш ш ш ш щ щ щ)

    # Vowels, normalized by character frequency
    VOWELS_NORMALIZED = %w(а а а а а а а а а а а а а а а а а а а а а а а а а а а а е е е е е е е е е е е е е е е е е е е е е е е е е е е е е е е и и и и и и и и и и и и и и и и и и и и и и о о о о о о о о о о о о о о о о о о о о о о о о о о о о о о о о о о о о о о о у у у у у у у у у ы ы ы ы ы ы э ю ю я я я я я я я)

    # Returns a random word. Passing a previous word affects the first character
    def self.word(previous_word=nil)
      if previous_word
        last_char = previous_word[-2,2]
        if CONSONANTS.include?(last_char)
          next_char = VOWELS_NORMALIZED
        else
          next_char = CONSONANTS_NORMALIZED
        end
      else
        next_char = rand>0.5 ? CONSONANTS_NORMALIZED : VOWELS_NORMALIZED
      end
      word = ''
      random_length(WORD_LIMITS).times do
        word << next_char[rand next_char.length]
        next_char = next_char===CONSONANTS_NORMALIZED ? VOWELS_NORMALIZED : CONSONANTS_NORMALIZED
      end

      word
    end

    # Returns a random sentence
    def self.sentence
      last_word = nil
      sentence = []
      random_length(SENTENCE_LIMITS).times do
        sentence << word(last_word)
      end
      capitalize_ru(sentence.join(' ')) + '.'
    end

    # Returns a random paragraph
    def self.paragraph
      sentences = []
      random_length(PARA_LIMITS).times do
        sentences << sentence
      end
      sentences.join(' ')
    end

    private

    # Random value based on limits
    def self.random_length(limits)
      limits.first + rand(limits.last-limits.first+1)
    end

    if  RUBY_VERSION < '1.9'
      # Capitalizes a russian string. Works only with UTF8.
      def self.capitalize_ru(word)
        w = word.clone
        code = (w[1]-0b10000000) + ((w[0]-0b11000000)<<6)
        code -= 32
        w[1] = (code & 0b00111111) | 0b10000000
        w[0] = (code >> 6) | 0b11000000
        w
      end
    else
      def self.capitalize_ru(word)
        w = word.clone
        w[0] = (w[0].ord-32).chr(Encoding::UTF_8)
        w
      end
    end
  end
end


lib/vydumschik/address.rb
# encoding: utf-8

require 'yaml'

module Vydumschik

  # This module generates random addresses
  module Address

    # Maximum number of apartments in building
    MAX_APT = 100

    # Maximum number of buildings for different types of streets
    BUILDING_LIMITS = {:street => 50, :alley => 10, :avenue => 120, :square => 10, :other => 20}

    # Probability of building without apartments
    HOUSE_PROBABILITY = 0.2

    # Probability of building number having a  modifier
    MODIFIER_PROBABILITY = 0.05

    # Building number modifiers
    BUILDING_MODIFIERS = %w(а б в г д)

    # Returns a random street name
    def self.street
      data[:streets][rand data[:streets].length][:name]
    end

    # Returns a random building number
    def self.building(limit = BUILDING_LIMITS[:other])
      bld = (1+rand(limit)).to_s
      bld+='-'+BUILDING_MODIFIERS[rand BUILDING_MODIFIERS.length] if rand<MODIFIER_PROBABILITY
      bld
    end

    # Returns a random apartment number
    def self.apartment
      (1+rand(MAX_APT)).to_s
    end

    # Returns a random street address
    def self.street_address
      street = data[:streets][rand data[:streets].length]
      bld = building(BUILDING_LIMITS[street[:type]])
      if rand>0.5
        apt = (rand<HOUSE_PROBABILITY) ? bld : bld+'/'+apartment
      else
        apt = (rand<HOUSE_PROBABILITY) ? "д. #{bld}" : "д. #{bld}, кв. #{apartment}"
      end
      "#{street[:name]}, #{apt}"
    end

    private

    def self.data
      unless @data
        @data = YAML.load_file(File.expand_path('../../../data/addresses.yml', __FILE__))
        encode_data_to_utf8 if RUBY_VERSION >= '1.9'
      end
      @data
    end

    def self.encode_data_to_utf8
      @data[:streets].each {|s| s[:name].force_encoding('utf-8') }
    end
  end
end
